modules/camera/camera_win.cpp

```diff
@@ -64,35 +414,293 @@ CameraFeedWindows::~CameraFeedWindows() {
+			printf("first sample failed\n");
+		}
+	}
+
+	if (FAILED(hr)) {
+		printf("failed to activate feed\n");
+		if (pSource) {
+			pSource->Shutdown();
+
+			// NOTE: The source reader shuts down the media source
+			// by default, but we might not have gotten that far.
+		}
+		CloseDevice();
+	}
+
+	SafeRelease(&pSource);
+	SafeRelease(&pAttributes);
+	SafeRelease(&pType);
+
+	LeaveCriticalSection(&m_critsec);
+
+	if (hr == S_OK)
+		return true;
+	else
+		return false;
+};
+
+void CameraFeedWindows::deactivate_feed() {
+	printf("deactivate_feed\n");
+	CloseDevice();
 };
 
+void CameraFeedWindows::set_device(IMFActivate *p_device) {
+	// put some data in
+	UINT32 cam_str_len;
+	HRESULT hr = S_OK;
+
+	LPWSTR cam_str = 0;
+	pActivate = p_device;
+
+	hr = pActivate->GetStringLength(MF_DEVSOURCE_ATTRIBUTE_FRIENDLY_NAME, &cam_str_len);
+	cam_str = (wchar_t *)malloc(cam_str_len * sizeof(wchar_t) + 1);
+	hr = pActivate->GetString(MF_DEVSOURCE_ATTRIBUTE_FRIENDLY_NAME, cam_str, cam_str_len + 1, NULL);
+	// wprintf(L"%ws\n", cam_str);
+	name = cam_str;
+	free(cam_str);
+
+	position = CameraFeed::FEED_UNSPECIFIED;
+}
+
 //////////////////////////////////////////////////////////////////////////
 // CameraWindows - Subclass for our camera server on windows
 
-void CameraWindows::add_active_cameras(){
+// update_feeds on OSX version
+void CameraWindows::add_active_cameras() {
 	///@TODO scan through any active cameras and create CameraFeedWindows objects for them
+	printf("CameraWindows::add_active_cameras()\n");
+	HRESULT hr = S_OK;
+	ChooseDeviceParam param = { 0 };
+	UINT iDevice = 0;
+	IMFAttributes *pAttributes = NULL;
+
+	hr = MFCreateAttributes(&pAttributes, 1);
+	if (FAILED(hr)) {
+		goto done;
+	}
+
+	hr = pAttributes->SetGUID(
+			MF_DEVSOURCE_ATTRIBUTE_SOURCE_TYPE,
+			MF_DEVSOURCE_ATTRIBUTE_SOURCE_TYPE_VIDCAP_GUID);
+	if (FAILED(hr)) {
+		goto done;
+	}
+
+	hr = MFEnumDeviceSources(pAttributes, &param.ppDevices, &param.count);
+
+	// only for enumerating all devices
+
+	printf("%d cameras detected\n", param.count);
+	for (int i = 0; i < param.count; i++) {
+		UINT32 cam_str_len;
+		_Post_ _Notnull_ LPWSTR cam_str = 0;
+		param.ppDevices[i]->GetStringLength(MF_DEVSOURCE_ATTRIBUTE_FRIENDLY_NAME, &cam_str_len);
+		//printf("%d" , cam_str_len);
+		cam_str = (wchar_t *)malloc(cam_str_len * sizeof(wchar_t) + 1);
+		hr = param.ppDevices[i]->GetString(MF_DEVSOURCE_ATTRIBUTE_FRIENDLY_NAME, cam_str, cam_str_len + 1, NULL);
+		if (FAILED(hr)) {
+			wprintf(L"can't get string\n");
+			free(cam_str);
+			goto done;
+		}
+		wprintf(L"%d-%ws\n", i, cam_str);
+		free(cam_str);
+
+		Ref<CameraFeedWindows> newfeed;
+		newfeed.instance();
+		newfeed->set_device(param.ppDevices[i]);
+		add_feed(newfeed);
+	}
+
+done:
+	SafeRelease(&pAttributes);
 };
 
 CameraWindows::CameraWindows() {
 	// Find cameras active right now
+	HRESULT hr = MFStartup(MF_VERSION);
 	add_active_cameras();
 
 	// need to add something that will react to devices being connected/removed...
```

Comment:
Yes we should add logic that ensures we detect cameras being connected or disconnected to the PC.

Reply:
there is an existing mechanism - the one used by the joypad to detect hotplug hardware - using WM_DEVICECHANGE - but it seems it's going to touch a lot of other files so I'll leave it for some other time.